src/org/opendatakit/briefcase/ui/reused/source/RemoteServerDialog.java

```diff
@@ -21,39 +21,57 @@
 import java.util.ArrayList;
 import java.util.List;
 import java.util.function.Consumer;
+import javax.swing.SwingUtilities;
 import org.opendatakit.briefcase.reused.RemoteServer;
-import org.opendatakit.briefcase.reused.http.Http;
+import org.opendatakit.briefcase.reused.http.HttpException;
 import org.opendatakit.briefcase.reused.http.Response;
 
 public class RemoteServerDialog {
-  private final RemoteServerDialogForm form;
+  final RemoteServerDialogForm form;
   private final List<Consumer<RemoteServer>> onConnectCallbacks = new ArrayList<>();
 
-  private RemoteServerDialog(RemoteServerDialogForm form, Http http) {
+  private RemoteServerDialog(RemoteServerDialogForm form, RemoteServer.Test serverTester) {
     this.form = form;
 
     this.form.onConnect(server -> {
-      Response<Boolean> response = server.testPull(http);
-      if (response.isSuccess()) {
-        triggerConnect(server);
-        form.hideDialog();
-      } else {
-        String error = response.isRedirection()
-            ? "Redirection detected.\n\nPlease, review the connection parameters and try again."
-            : response.isUnauthorized()
-            ? "Wrong credentials.\n\nPlease, review the connection parameters and try again."
-            : response.isNotFound()
-            ? "Aggregate not found.\n\nPlease, review the connection parameters and try again."
-            : "Please, review the connection parameters and try again.";
-        showErrorDialog(form, error, "Wrong connection parameters");
-      }
+      form.setTestingConnection();
+      new Thread(() -> SwingUtilities.invokeLater(() -> {
```

Comment:
Ugh, it's ugly, but that gets the best UX. Without the new Thread, the progress bar doesn't start the animation and without the `invokeLater` you're not doing UI changes "the Swing way" (updating the UI outside the EDT).

Reply:
Why not do the server connection in a “worker” thread, and only do `showError` and `hideDialog` on the EDT? If no host is answering or there is some other delay, doesn’t this hang the UI?

Reply:
I’ll write something quickly using SwingWorker, which has a done method called on the EDT.

Reply:
Since most of the code is about UI side-effects (there are 4 UI side-effects vs 1 HTTP side-effect), It felt cleaner to just wrap everything inside the EDT.

Also, I need to evaluate the result of the HTTP request to decide what to do or manage exceptions... I don't see how it could be done, although what you say sounds better. Could you maybe write your version to compare?

Reply:
SwingWorker is perfect for this.

https://github.com/dcbriccetti/briefcase/commit/645e1fb92502befdd085837d3b147d61b497e16c

Reply:
Later, maybe let’s come back and add cancel support to the dialog, and lengthen the timeout for host connection.